Backend commits for mute users feature.

abhijeetbodas2001 opened the pull request:
This PR contains only the backend commits from #16915, separated out for convenience.
I've addressed the review comments on these commits on the main PR, and am writing the changes here-

- https://github.com/zulip/zulip/pull/16915#discussion_r608292098
    > I guess I didn't update this commit for the muted_user_profile => muted_user rename; can you take care of that?

    Done.

- https://github.com/zulip/zulip/pull/16915#discussion_r608292553

    > Can you get the raw SQL for this query and do an EXPLAIN ANALYZE on it in a manage.py dbshell (postgres shell) to make sure the query plan is an index scan? You should be able to grab the generated SQL via the queries_captured context manager and just printing the queries rather than checking the length.

    Its a `Bitmap Index Scan`
```sql
                                                                   QUERY PLAN
------------------------------------------------------------------------------------------------------------------------------------------------
 Bitmap Heap Scan on zerver_muteduser  (cost=4.21..14.37 rows=8 width=4) (actual time=0.010..0.011 rows=0 loops=1)
   Recheck Cond: (muted_user_id = 8)
   ->  Bitmap Index Scan on zerver_muteduser_muted_user_id_d4b66dff  (cost=0.00..4.21 rows=8 width=0) (actual time=0.006..0.006 rows=0 loops=1)
         Index Cond: (muted_user_id = 8)
 Planning time: 0.120 ms
 Execution time: 0.049 ms
(6 rows)
```


- https://github.com/zulip/zulip/pull/16915#discussion_r608293110

    > You can make this comment a docstring, and that'd probably be nicer. Also, add another sentence mentioning that this will include deactivated users. (I think that behavior is probably correct, given how we'll use this).

    Done.

- https://github.com/zulip/zulip/pull/16915#discussion_r608293469

    > Why AbstractSet, and not Set?

    Copy-paste error from my side, fixed.

- https://github.com/zulip/zulip/pull/16915#discussion_r608293710

    > Let's call this test_new_messages... -- aka add new

    Done!

- https://github.com/zulip/zulip/pull/16915#discussion_r608293878

    > do_mark_muted_user_messages_as_read would be a better name.

    Makes sense, done.


- https://github.com/zulip/zulip/pull/16915#discussion_r608294122

    > I'd prefer that these first several lines were a single block rather than reassigning the variable several times; also, we should call the variable messages -- our naming convention is to avoid abbreviations like msgs.

    Ok, done.


- https://github.com/zulip/zulip/pull/16915#discussion_r608294833

    > It would be a nice preparatory cleanup to create a dataclass for the update_message_flags event type so we stop duplicating these dictionaries, but would also be a fine followup, so feel free to just copy this idea onto your TODO list and focus on the rest.

    Will do this as a followup.


- https://github.com/zulip/zulip/pull/16915#discussion_r608295103

    > Add a comment before this line explaining what API call we're about to do / what we're about to test.

    Added a comment, and I also split this into a separate test, which I think improves readability.

- https://github.com/zulip/zulip/pull/16915#discussion_r608295288

    > Let's make do_mute_user call this internally? That's safer, since it handles the case where some other code path starts being able to mute users as well.

    Done. This also forced me to expand the events test to also test that the `read` event is sent when a user is muted 😄

timabbott commented on zerver/tests/test_muting_users.py:
```diff
@@ -220,3 +220,33 @@ def test_new_messages_from_muted_user_marked_as_read(self) -> None:
+        # Then test that the messages sent earlier got marked as read.
+        stream_message = self.send_stream_message(cordelia, "general", "Spam in stream")
+        huddle_message = self.send_huddle_message(cordelia, [hamlet, iago], "Spam in huddle")
+        pm_to_hamlet = self.send_personal_message(cordelia, hamlet, "Spam in PM")
```
This test should verify the messages are unread before the operation.  Possibly makes sense to factor out a function we can call twice, maybe `assert_usermessages_read_flag(messages, True|False)` to avoid messy duplication.

timabbott commented:
This is great; posted one comment above (and I also intend to tighten the big block comment and/or move parts of it to `docs/subsystems/sending-messages.md` before merging).

abhijeetbodas2001 commented:
I have addressed https://github.com/zulip/zulip/pull/18086#discussion_r612033498. I extracted out a lot of stuff in a helper, and asserted that the messages are unread before muting. And I also added some assertions for the "other" user (`Othello`). I believe these will make it even more clear what we're testing.

I haven't touched the comment though. Not sure if they belong in `sending-messages.md`, since they're more about fetching messages?

timabbott commented:
Yeah, I wanted to do an edit pass on the comments for brevity/location/clarity; I've done so and then merged this.

Thanks @abhijeetbodas2001!
